beatmap_reader/hitobject/std/std_holdnote_hitobject_base.py

```diff
@@ -31,46 +35,65 @@ def __init__(self, **kargs):
-            self.tdata.append([ x_pos, y_pos, beat_time ])
+        if self.end_time() == self.start_time():
+            pos = self.hdata[Hitobject.HDATA_POSX], self.hdata[Hitobject.HDATA_POSY]
+            self.tdata.append([ *pos, self.start_time() ])
+            return
+
+        velocity = kargs['velocity']
+        ms_per_beat = kargs['beat_length'] / kargs['tick_rate']
+        ms_per_repeat = self.px_len / velocity
+
+        tick_times = list(frange(self.start_time() + ms_per_beat, self.start_time() + ms_per_repeat, ms_per_beat))
+        # https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/Objects/SliderEventGenerator.cs#L24
+        while len(tick_times) > 0 and self.__time_to_dist(tick_times[-1]) > self.px_len - 10 * velocity:
+            tick_times.pop()
+        ticks = [ (self.time_to_pos(tick_time), tick_time - self.start_time()) for tick_time in tick_times ]
+
+        # https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/Objects/SliderEventGenerator.cs#L115
+        for repeat in range(self.repeats):
+            reverse = repeat % 2 == 1
+
+            repeat_start_time = self.start_time() + repeat * ms_per_repeat
+            x_pos, y_pos = self.gen_points[-1] if reverse else self.gen_points[0]
+            self.tdata.append([ x_pos, y_pos, repeat_start_time ])
+
+            if reverse:
+                self.tdata.extend([ *pos, repeat_start_time + (ms_per_repeat - time) ] for pos, time in reversed(ticks))
+            else:
+                self.tdata.extend([ *pos, repeat_start_time + time ] for pos, time in ticks)
 
-        if self.tdata[-1][Hitobject.TDATA_T] != self.end_time():
-            x_pos, y_pos = self.time_to_pos(self.end_time())
-            self.tdata.append([ x_pos, y_pos, self.end_time() ])
+        # https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/Objects/SliderEventGenerator.cs#L79
+        # https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/Objects/Legacy/ConvertSlider.cs#L52
+        midpoint_time = (self.start_time() + self.end_time()) / 2
+        end_tick_time = max(self.end_time() - 36, midpoint_time)
+        x_pos, y_pos = self.time_to_pos(end_tick_time)
+        self.tdata.append([ x_pos, y_pos, end_tick_time ])
 
 
     def time_to_pos(self, time):
-        return self.__idx_to_pos(self.__percent_to_idx(value_to_percent(self.start_time(), self.end_time(), time)))
+        return self.__dist_to_pos(self.__time_to_dist(time))
 
 
-    # TODO: make sure this is correct
-    # TODO: test a slider 200px across with various repeat times and tick spacings
-    def get_velocity(self):
-        return self.px_len / (self.end_time() - self.start_time())
-
-
-    def __idx_to_pos(self, idx):
-        if idx > len(self.gen_points) - 2:
-            return self.gen_points[-1]
+    def __time_to_dist(self, time):
+        percent = value_to_percent(self.start_time(), self.end_time(), time)
+        if percent < 0.0: return 0
+        if percent > 1.0: return self.px_len
```

Comment:
Why not `percent <= 0.0` and `percent >= 1.0`?

Reply:
I initially left [those](https://github.com/abraker-osu/osu_beatmap_reader/blob/efc4a769eac0351896e44209cb2d8b483e3e2648/beatmap_reader/hitobject/std/std_holdnote_hitobject_base.py#L67-L68) checks as-is because I didn't want to change more code than necessary, but I'll remove them since `value_to_percent` uses `bound` anyway.